### PR TITLE
fixed https://github.com/wso2/carbon-auth/issues/29

### LINKED ISSUE #29: Key Manager/UserStore - LDAP User Store Connector
Component - Key Manager/UserStore 

LDAP User Store Connector

### PR DESCRIPTION
**Pull request by ruks**
fixed https://github.com/wso2/carbon-auth/issues/29
https://github.com/wso2/carbon-auth/issues/90
https://github.com/wso2/carbon-auth/issues/91
https://github.com/wso2/carbon-auth/issues/92
https://github.com/wso2/carbon-auth/issues/93

reduced org.wso2.carbon.auth.user.store required code coverage to 60% and exclude org.wso2.carbon.auth.user.store.connector.ldap.LDAPUserStoreConnectorIT.java class since it is covers by Integration testing.

to run integration testing use,
 mvn clean install -P local-ldap

### CODE REVIEW

**Comment by ruks** [comment]
draft doc: https://docs.google.com/document/d/1tlVHHwv_YDa4gZTEEmSNYfoJlSs7oIOD4NMK432vXHA/edit?usp=sharing

File: components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import org.wso2.carbon.auth.user.store.connector.UserStoreConnectorFactory;
+import org.wso2.carbon.auth.user.store.connector.jdbc.DefaultPasswordHandler;
+import org.wso2.carbon.auth.user.store.constant.UserStoreConstants;
+import org.wso2.carbon.auth.user.store.exception.UserNotFoundException;
+import org.wso2.carbon.auth.user.store.exception.UserStoreConnectorException;
+
+import java.security.NoSuchAlgorithmException;
+import java.util.Map;
+
+//import java.util.HashMap;
```
Comment (harsha89) [inline components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java]: remove unwanted imports

File: components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        this.userStoreConnector = userStoreConnector;
+    }
+
+    @Override
+    public boolean doAuthenticate(String userName, Object credential) throws UserStoreException {
+        try {
+            String password = (String) credential;
+            String userId = userStoreConnector.getConnectorUserId(UserStoreConstants.CLAIM_USERNAME, userName);
+            Map info = userStoreConnector.getUserPasswordInfo(userId);
+
```
Comment (harsha89) [inline components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java]: do we require these spaces

File: components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java [hunk 3]
```diff
@@ -70,23 +64,40 @@ public DataSource getDataSource(String dataSourceName) throws DataSourceExceptio
...
         }
         List<Attribute> attributeList = new ArrayList<>();
         attributeList.add(new Attribute(UserStoreConstants.CLAIM_USERNAME, user));
+        try {
+            String uid = connector.getConnectorUserId(UserStoreConstants.CLAIM_USERNAME, user);
+            if (uid != null) {
+                log.info("default user already exist");
+                return;
+            }
+        } catch (UserNotFoundException e) {
```
Comment (harsha89) [inline components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java]: log with e?

Comment (harsha89) [reply components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java]: should these info logs convert to debug logs?

File: components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            if (hashedPassword.equals(info.get(UserStoreConstants.PASSWORD))) {
+                return true;
+            } else {
+                return false;
+            }
+
+        } catch (UserStoreConnectorException e) {
+            throw new UserStoreException("User Connector exception occurred", e);
+        } catch (UserNotFoundException e) {
+            throw new UserStoreException("User not found exception occurred", e);
```
Comment (harsha89) [inline components/auth/org.wso2.carbon.auth.user.mgt/src/main/java/org/wso2/carbon/auth/user/mgt/impl/LDAPUserStoreManager.java]: dont we need to log and throw the exception?

File: components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java (hunk 3 above)
Comment (harsha89) [reply components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java]: log and throw the exception. I don't think this user store manger use only in out rest api layer

File: components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/connector/ldap/LDAPConnectionContext.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                //                context = (DirContext) new LdapCtxFactory().getInitialContext(environment);
+                context = LdapCtxFactory.getLdapCtxInstance(environment.get(Context.PROVIDER_URL), environment);
+                return context;
+            } catch (NamingException e) {
+                log.error("Error obtaining connection. " + e.getMessage(), e);
+                if (retry == 1) {
+                    throw new LDAPConnectorException("Error obtaining connection. " + e.getMessage(), e);
+                }
+            }
+
```
Comment (harsha89) [inline components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/connector/ldap/LDAPConnectionContext.java]: spaces needed?

**Review by harsha89 — CHANGES_REQUESTED** [review]

File: components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java (hunk 3 above)
Comment (ruks) [reply components/auth/org.wso2.carbon.auth.user.store/src/main/java/org/wso2/carbon/auth/user/store/internal/ConnectorDataHolder.java]: this is temp code to add default user to user store. I think this user need to be added using SCIM

**Review by harsha89 — APPROVED** [review]

**Pull request merged**